### ISSUE TITLE
README should document JNA dependency versioning use cases

### ISSUE THREAD

**Issue by AdamQwy** [oshi/oshi#1040]
java.lang.NoClassDefFoundError: com/sun/jna/platform/win32/VersionHelpers

**Comment by AdamQwy** [comment]
this is what i dependence in my project pom file
`
        <dependency>
            <groupId>com.github.oshi</groupId>
            <artifactId>oshi-core</artifactId>
            <version>4.1.1</version>
        </dependency>
`

i finally found 
when I add the dependency of 4.1.1 , 
the version of jna is not follow the "oshi-core" 's version.

so i add this into tag \<properties\> 👇
`
<jna.version>5.4.0</jna.version>
`
then it worked correctly.

😊Hoping to add this into README,  it may confuse other guys.

**Comment by dbwiddis** [comment]
That definition is already in `oshi-core`'s POM [here](https://github.com/oshi/oshi/blob/oshi-parent-4.1.1/oshi-core/pom.xml#L49).  Are you manually defining another version?  Do you have another dependency in your project that uses an earlier version of JNA?

I'm not sure exactly what problem you solved in order to document it. :)

**Comment by arjhun** [comment]
I found out that the <jna.version>4.5.2</jna.version> property is set by spring-boot-starter-parent.  Moving up the Oshi deppendency up the tree like mentioned in the README doens't effect the version being used. If you are using spring boot starter, manually specify the jna version for oshi!

**Comment by dbwiddis** [comment]
Hmm, that's interesting.  So it's not the maven dependency resolution that's the issue here, it's the fact that a property is defined earlier?   Seems that we could change the property name in oshi's POM to avoid this conflict with Spring Boot....

**Comment by arjhun** [comment]
Yes That could help, I think that maven will throw an error though... maybe use the enforcer plugin? https://maven.apache.org/enforcer/maven-enforcer-plugin/ This will add another dependency. I have to say this is where my knowledge ends... I could dive into this if needed. Thanks for your amazing project!

**Comment by hazendaz** [comment]
Every dependency in spring boot starter parent can be overridden.  However it doesnt care how this oshi pom is setup. The suggested fix for spring boot parent is correct.  We just need to point that out on readme.  For those that use spring boot dependencies instead the order is what works.

Enforcer plugin wont provide anything useful from oshi standpoint.  This has to do with how mavens reactor works and what boot is trying to leverage for strict enforcement.  The order in the chain matters and they use dependency management.  The jna libraries could be put into dependency management but the influence there is unlikely to just win out as the starter parent gets the priority.

Note this same issue affects waffle and we have direct spring boot starters.  The latest jna binaries also simply work so risk is low.

Unfortunately this is a dance one must play if using springs starter parent to control all.  As spring states they are very opinionated and their parent use is not required.  They have the two options.  Both are fine though but one must understand springs use of maven is not magic and it's not the be all end all to managing dependencies.  Care must still be taken by users in the end.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by dbwiddis** [comment]
Thanks for the comments.  I do think that updating the `README` is the right answer, but I want to be clear on exactly what is causing the issue.  Is a user using the spring boot parent causing all those dependencies to process first?   Or is using the spring boot parent setting the `jna.version` property which can be overridden in the user's POM but not OSHI's POM?

Is the correct fix:
 - User must manually specify the property `jna.version`
 - If OSHI renames its dependency property (e.g., `oshi.jna.version`) will that dependency process in priority order to the Spring Boot dependency?  (It's possible the answer is no, just wanting to know this.)

**Comment by hazendaz** [comment]
OSHI pom use of properties has no impact to the reactor.  It's the parent to child relationship.  With spring boot parent, it defines properties for the dependencies it contains in dependency management so one can override them.  The child properties will take presidence over the parent property values.  It does not look into other dependency poms and their respective property definitions.  The only real influence you have there is to move 'jna' into dependency management.  That will have some but limited influence since spring is doing exactly that both in the spring boot parent and the spring boot dependencies.  Thus you have to override the version with spring boot parent and you have to force ordering with spring boot dependencies.  Normally I don't think we would document such information for a specific library or framework.  However, spring boot is so big at this point it sort of demands that we do so.

**Comment by hazendaz** [comment]
Another option is to somehow get spring boot to want to use oshi for some reason ;)

**Comment by hazendaz** [comment]
spring boot actuator might be some avenue into spring boot...

**Comment by dbwiddis** [comment]
> With spring boot parent, it defines properties for the dependencies it contains in dependency management so one can override them. 

Right, this is why redefining the property works so that spring boot loads the correct dependency, and I can easily document this in the README.  My bigger question is, what happens when OSHI is listed in someone's POM as the first dependency?  Does using a parent pom mean that all of the parent's dependencies (which become the calling pom's dependencies) are processed by Maven's dependency manager first, before individually listed ones in the POM?  Or are they processed second... in which case it still fails because while oshi itself would be first, it would hit the JNA dependencies from the parent at the first level as well, before processing oshi's second-level dependencies.  

> Another option is to somehow get spring boot to want to use oshi for some reason ;)

Heh!

**Comment by hazendaz** [comment]
Maybe the right thing for now is to move jna to dependency management in oshi assuming it is not now.  Then have one of the users here try a new release to see what behaviour diference that actually causes.  I think for the sake of it, I'm going to move waffle jna usage into dependency management as I'm getting ready to release over there.  I'm also going to add to readme about overriding jna.  It's also pretty easy to put together a quick hello world with spring boot trying both their ways if we want to confirm how it works.

In general though, because the spring boot parent defines the dependencies it needs, it comes first.  And actually their starter parent is just daisy chained up to the same dependencies module one can use instead of their parent.  See https://search.maven.org/classic/remotecontent?filepath=org/springframework/boot/spring-boot-dependencies/2.2.4.RELEASE/spring-boot-dependencies-2.2.4.RELEASE.pom.  They don't specifically call out the direct usage in this.  Instead they put it in dependency management which locks down that version so if something else calls it, its likely to stick to their version only.  Due to its position in the grand scheme of things it will win out.  In a parent to child, it's always only resolvable to what we want by defining the property.  That makes the effective pom have the right version.  So that one is easy to understand.  We set the version, which by their design applies that in dependency management and it gains influence over all others.

The ordering of 'oshi' in other use cases then needs to come before spring dependencies module.  That way it forces up the usage.  But to be safe there, having it also listed in dependency management will put more pressure on maven to ensure we win out.

So in oshi the jna usage would be defined twice.  One in dependency management then later when used.  It is possible the ordering then might not matter but that would need tested to confirm out.

So I would approach this as first getting this into dependency management since it's the most critical part of oshi (same with waffle).  Then perform a couple simple tests, one using spring boot starter parent and doing nothing to see what resolves out.  And second to just call their dependencies in import scope (them first, then us), to see what resolves out.  Then adjust that order.  That will ultimately lead to what is the right thing to note in the README when it comes to spring boot.

**Comment by dbwiddis** [comment]
> Maybe the right thing for now is to move jna to dependency management in oshi assuming it is not now. 

Agreed this is a good plan to help with many other use cases, although since JNA is listed in the Spring Boot `dependencyManagement` section I think that will take precedence over any possible thing we can do in OSHI's POM.  

So my thoughts are, as immediate steps:
 - In the `README`, document the `jna.version` property fix for Spring Boot, with a comment that usage of any other parent may require a similar property override.
 - In the `oshi-core` `pom.xml` file, put the JNA dependencies into `dependencyManagement`.

Not sure what follow on steps are possible, but I don't see a reason to wait on the above before testing.
